### ISSUE TITLE
Web player cuts off tracks based on expected rather than actual data size

### ISSUE THREAD

**Issue by eharris** [airsonic/airsonic#723]
## Problem description

In my testing of a workaround for doing automatic replaygain via transcoding (see issue #525), I discovered that the web app seems to be cutting off tracks early despite there being more data to play.  This was tested and confirmed both on Firefox (Debian jessie 52.7.2esr-1~deb9u1) and Chrome (Version 64.0.3282.186 (Official Build) (64-bit) on OS X).

Testing with an actual Subsonic player that uses the API does not exhibit the problem.  Only the web player seems to be affected.

### Steps to reproduce

1. Change your user to have a bitrate limit of 32kbps in the user settings.
2. Change the downsample command on the transcoding page to replace the variable `%bk` with a fixed higher bitrate, like `128k`.  If you were previously using the stock ffmpeg transcode, the new command line for downsample would be `ffmpeg -i %s -map 0:0 -b:a 128k -v 0 -f mp3 -`
3. Use the web interface to play an mp3 file that has a bitrate higher than the limit configured in step 1 (so that it will get transcoded).
4. Notice that with the given values, the song will only play for 1/4 of the time (32k/128k), and the duration shown in the web player is also 1/4 of the actual song duration.  If you use 32k and 64k, the song will be cutoff 1/2 way in.

## System information

 * **Airsonic version**: 10.1.1-RELEASE – December 16, 2017
 * **Operating system**: Debian Linux (jessie)
 * **Proxy server**: None
 * **Client**: Firefox (Debian jessie 52.7.2esr-1~deb9u1) and Chrome (Version 64.0.3282.186 (Official Build) (64-bit) on OS X).

## Additional notes

I was using this config (an unusually low bitrate limit) in order to force airsonic to always transcode the files but only for particular users so that I could use the transcoding callout to apply replay gain to the audio.  Again, see #525 for more details on why.  I don't know of another way to force certain specific users playback to always go through a transcode.  Is there another way?

**Comment by ae-g-i-s** [comment]
I'm not completely sure, but this is probably related, unless somebody fixed this issue in Airsonic: https://github.com/tsquillario/Jamstash/issues/131#issuecomment-349507059

The server will sometimes send an incorrect Content-Range header and simply refuse to deliver the remaining bytes (because they don't exist). DSub for Android doesn't care, and neither does my hacky bash client, because they just grab the entire file and don't request a specific range. Browsers do care because they try to be smart about chunked transfers.

**Comment by eharris** [comment]
It seems like it's a problem that the server is responding with a Content-Length at all for the streaming of the output of any transcoding process.  Since the server doesn't know until the transcode is complete, it shouldn't specify the length rather than specifying a length that is almost certainly wrong.  This seems like a clear bug for any transcoded source.

**Comment by KingDuckZ** [comment]
I'm wondering if that could be the cause of my mp3/flac playback problem as well. In version 10.1.1 playback of such files produces 2 seconds of music and the playback bar marks 00:02 as the total length, which is an upright lie. In the meantime Airsonic's log gets spammed with a broken pipe error followed by a very verbose java backtrace, but it seems to be retrying whereas the browser player just gives up and moves on to the next song.

**Comment by tari** [comment]
Duplicate of #685 I think, where I've been exploring possible solutions.

**Comment by RuralHunter** [comment]
I see the same issue. I checked the source and think the problem is because the bitrate set in transcoding command is not used when calculating the content length.

**Comment by ae-g-i-s** [comment]
> I checked the source and think the problem is because the bitrate set in transcoding command is not used when calculating the content length.

It is impossible to determine the Content-Length based on the bitrate because of headers, VBR (which all modern audio codecs use), tags/comments etc. See #685.

**Comment by RuralHunter** [comment]
Yes, but this problem is not about how to calculate the content length. This problem is that, the fixed bitrate set in trancoding command is not extracted at all. So with current code base, the bitrate should always be set in user settings.

**Comment by ae-g-i-s** [comment]
I'm not sure why you would need to know the bitrate, could you elaborate?

It is perfectly valid for a transcode command to hardcode the bitrate in some arbitrary format or use non-bitrate options (like `-q` for `oggenc`, `-8` for `flac`). Any effort to parse this in airsonic is futile because it can not know the command line options for all encoders (in fact, it can not even determine what the encoder is). Also, VBR/ABR means that the bitrate option is almost irrelevant.

**Comment by RuralHunter** [comment]
To determine the content length, you need to know the bitrate used in trancoding. If you use a fixed value in transcoding command which is different than the bitrate in user setting, there is no way to calculate the correct content length as you also agreed that parsing the transcoding command is not a valid solution.

**Comment by ae-g-i-s** [comment]
You can not determine the content length based on the bitrate, though - see above ;)

**Comment by RuralHunter** [comment]
well, I maybe mis-understood. but what's your point? It's possible to determine the content length without bitrate?

**Comment by ae-g-i-s** [comment]
No, it's impossible to determine the content length for transcoded files, period. Doesn't matter if you know the bitrate.
Chunked content encoding without a known content length is most likely the correct way to go. https://github.com/airsonic/airsonic/pull/798 mitigates this by *not* setting the `Accept-Ranges` header by default, so clients don't try to seek by requesting a byte range.

See, for instance, this response from nginx on a related project that uses CGI to transcode a file:
```
HTTP/1.1 200 OK
Server: nginx/1.14.1
Date: Thu, 10 Jan 2019 08:26:03 GMT
Content-Type: audio/ogg
Transfer-Encoding: chunked
Connection: keep-alive
```
After this, the server just sends the entire content. No Content-Length needed. This situation is very similar to the one airsonic is facing: The server can not know how much data a CGI script or FastCGI server will send. It could theoretically try to buffer the entire response and then just count the number of bytes, but nothing guarantees that CGI won't produce multiple GB of data, so that's not a sane option.

The very short version: **Content-Length is not required** (as long as you use Transfer-Encoding: chunked). And airsonic can't possibly provide a sane value for Content-Length.

**Comment by RuralHunter** [comment]
OK, I understand. My point is the current code base uses bitrate to determine the content length, and then the web client side uses it to set the length of the music. I can see the server side set another head X-Content-Duration but don't know some how the web client doesn't use it.

**Comment by uqs** [comment]
/me too!
This was still working in the webplayer at least a year ago or so (I use it rarely). Playing the files via the Subsonic Android app also works, it will download the full file no matter what.

The webplayer however will only play back 2 seconds of a song (and claim they are all about 2-3s in length).

So can Content-Length be removed altogether?

**Comment by eharris** [comment]
There have been some changes recently regarding reworking of the stream controller code to use chunked encoding and not send content length or allow range requests.  Unfortunately, this has some very bad side effects (see #1117) due to the web player not allowing seeking even within buffered portions of a chunked stream, and undesired increased bandwidth usage in mobile or other situations where partial transfers and resuming are needed to minimize bandwidth use.

I've tried to address these problems while still allowing for range requests and predicted content length in #1123, but by an improved means and only in safer situations.

**Comment by eharris** [comment]
While I think the underlying cause of the original problem this ticket was meant to address has been "fixed" (as @tari noted by #685) by not allowing transcoded streams to have their length estimated, that has some really bad side effects that I've noted above, so I don't really consider it a long-term solution.

Since we still need a better method of addressing this problem **without** causing those other side effects, I'm going to let this stay open in hopes of a better solution.

**Comment by hawken93** [comment]
Bump! I think this might be my issue, I am transcoding flac to ogg using this
`ffmpeg -i %s -map 0:a -b:a %bk -v 0 -c:a libopus -f ogg -`
and the web player works horribly, it does like 15 seconds and then gives up. Which is kind of dissimilar, but I still think it has to be the same issue. 
"stream output exceeded expected length of 5904000. It is likely that the transcoder is not adhering to the bitrate limit or the media source is corrupted or has grown larger"

I tried to find out what kind of containers that subsonic was supposed to work with, and I see nothing but mp3 recommended. For the server to know exactly how many bytes the stream is going to be, means it really needs a good understanding of the container, and if this just works for mp3 and then in some cases - then that could explain the thought process..

I can think of a couple of ways to deal with this:
a) by finishing the transcode before serving the file, with due consequences. Maybe server side caching would have to be improved for this.
b) by supporting only specific containers and then reading and analyzing them during streaming to figure out sizes from that..
c) Go for HLS.. I think those formats were created just because of this problem..

**Comment by tari** [comment]
Yeah, the bitrate estimation assumes true CBR so using better codecs is difficult as a result. Opus prefers to operate in a VBR mode, so you may have better luck forcing it to CBR: [`ffmpeg -c:a libopus -vbr off`](https://ffmpeg.org/ffmpeg-codecs.html#libopus-1).

HLS/DASH are somewhat difficult because you need to determine how to chunk the output and dynamically generate a manifest- it's doable but a larger project. I would be happy with preset ffmpeg options for transcoding where you might select your desired codec and bitrate without writing an ffmpeg command line (and an "advanced" option to write your own) too, but again that seems a bit complex.

DASH or HLS would be ideal but is probably the most work; either improvement seems like it just needs somebody with enough motivation to implement it.

**Comment by hawken93** [comment]
Thanks for the feedback.

One takeaway I have is that I really want the option back to force off the content-length math because airsonic doesn't know better and the heuristic sucks. As you can see, I get VBR even though I have that %b in my cmdline.

I like the idea to present options about ffmpeg codecs and stuff. I think it's good because it allows you to remove containers known to not work at all, and make life easier.

As for DASH/HLS.. Where do I even start.. I don't think I'll be attempting it today at least. But what is clear to me is that this content length mismatch thing ruins the ability to have consistency from the ffmpeg output to the client, so the files being consistent enough is left to chance. Gapless playback is somewhat questionable - I think as long as the appended stuff is junk, the player will skip past it and buffer ahead so there is no problem. Ideas to insert silent samples should be shot down, but maybe the junk should be engineered so that it does not resemble anything that a player would detect as valid data. But this is all a workaround.

I would even be happy with an option to transcode before sending to client, to at least try and see how it feels. Combined with storing it over time, the server can build up a library of pretranscoded files. For busy servers this can save a lot of compute and be much faster. I do see the issue with weak servers and long files. Which is why it should be configurable. The advanced settings tab is way too short ;)
